File: nagasena/fixture/org/openexi/scomp/EXISchemaFactoryTestUtil.java
package org.openexi.scomp;

import java.io.*;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import org.xml.sax.*;

import org.openexi.schema.EXISchema;

public class EXISchemaFactoryTestUtil {

  public static EXISchema getEXISchema(Class<?> cls, EXISchemaFactoryErrorHandler compilerErrorHandler)
    throws Exception {
    return getEXISchema((String)null, cls, compilerErrorHandler);
  }

  /**
   * Loads schema then compiles it into EXISchema.
   * Schema file is resolved relative to the specified class.
   */
  public static EXISchema getEXISchema(String fileName, Class<?> cls,
      EXISchemaFactoryErrorHandler compilerErrorHandler)
      throws Exception {
    return getEXISchema(fileName != null ? new String[] { fileName } : new String[0],
        cls, new EXISchemaFactoryTestUtilContext(compilerErrorHandler));
  }

  public static EXISchema getEXISchema(String fileName, Class<?> cls,
      EXISchemaFactoryTestUtilContext context) throws Exception {
    return getEXISchema(fileName != null ? new String[] { fileName } : new String[0],
        cls, context);
  }

  static EXISchema getEXISchema(String[] fileNames, Class<?> cls,
      EXISchemaFactoryTestUtilContext context)
          throws Exception {
    EXISchemaFactory schemaCompiler = new EXISchemaFactory();
    schemaCompiler.setCompilerErrorHandler(context.errorHandler);
    schemaCompiler.setEntityResolver(context.entityResolver);
    
    URL url = null;
    final InputSource[] inputSources = new InputSource[fileNames.length];
    for (int i = 0; i < fileNames.length; i++) {
      final String fileName = fileNames[i];
      if ((url = cls.getResource(fileName)) != null) {
        InputSource inputSource = new InputSource(url.openStream());
        inputSource.setSystemId(url.toString());
        inputSources[i] = inputSource;
      }
      else
        throw new RuntimeException("File '" + fileName + "' not found.");
    }
    if (url == null)
      url = cls.getResource("/file.txt");
    
    EXISchema compiled;
    if ((compiled = schemaCompiler.compile(inputSources)) != null) {
      //saveXsc(compiled, url);
      final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
      new EXISchemaWriter().serialize(compiled, byteArrayOutputStream, context.stringBuilder);
      final byte[] grammarBytes = byteArrayOutputStream.toByteArray();
      //saveGram(grammarBytes, url);
      compiled = readEXIGrammar(grammarBytes, context.schemaReader);
      InputStream serialized = serializeSchema(compiled);
      return loadSchema(serialized);
    }
    return null;
  }
  
  @SuppressWarnings("unused")
  private static void saveXsc(EXISchema schema, URL url) throws Exception {
    if (url != null) {
      String urlString = url.toString();
      urlString = urlString.substring(0, urlString.lastIndexOf('.')) + ".xsc";
      String filePath = new URL(urlString).getFile();
      DataOutputStream out = new DataOutputStream(new FileOutputStream(filePath));
      schema.writeOut(out);
      out.close();
    }
  }

  @SuppressWarnings("unused")
  private static void saveGram(byte[] grammar, URL url) throws Exception {
    if (url != null) {
      String urlString = url.toString();
      urlString = urlString.substring(0, urlString.lastIndexOf('.')) + ".gram";
      String filePath = new URL(urlString).getFile();
      FileOutputStream out = new FileOutputStream(filePath);
      out.write(grammar);
      out.close();
    }
  }

  private static EXISchema loadSchema(InputStream is) throws IOException, ClassNotFoundException {
    EXISchema schema = null;
    DataInputStream dis = new DataInputStream(is);
    try {
      schema = EXISchema.readIn(dis);
    }
    finally {
      if (dis != null) dis.close();
    }
    return schema;
  }
  
  private static InputStream serializeSchema(EXISchema schema) throws IOException {
    InputStream is = null;
    ByteArrayOutputStream bos = new ByteArrayOutputStream();
    DataOutputStream dos = new DataOutputStream(bos);
    try {
      schema.writeOut(dos);
      dos.flush();
      bos.flush();
      is = new ByteArrayInputStream(bos.toByteArray());
    }
    finally {
      dos.close();
      bos.close();
    }
    return is;
  }
  
  /**
   * Serialize bytes to a file.
   * @param bts
   * @param baseFileName a file that exists in the destination
   * @param fileName name of the file the bytes are output to
   * @param cls
   */
  public static void serializeBytes(byte[] bts, String baseFileName, String fileName, Class<?> cls) 
     throws IOException, URISyntaxException {
    URI uri = cls.getResource(baseFileName).toURI();
    uri = uri.resolve(fileName);
    
    FileOutputStream fos;
    fos = new FileOutputStream(uri.toURL().getFile());
    
    fos.write(bts);
    fos.close();
  }
  
  private static EXISchema readEXIGrammar(byte[] grammarBytes, EXISchemaReader schemaReader) throws Exception {
    final InputStream inputStream;
    inputStream = new ByteArrayInputStream(grammarBytes);
    if (schemaReader == null)
      schemaReader = new EXISchemaReader();
    return schemaReader.parse(inputStream);
  }
  
}

File: nagasenacs/fixture/Nagasena/Scomp/EXISchemaFactoryTestUtil.cs
using System;
using System.IO;
using System.Text;

using Org.System.Xml.Sax;

using HeaderOptionsOutputType = Nagasena.Proc.HeaderOptionsOutputType;
using SchemaId = Nagasena.Proc.Common.SchemaId;
using Transmogrifier = Nagasena.Sax.Transmogrifier;
using EXISchema = Nagasena.Schema.EXISchema;
using EmptySchema = Nagasena.Schema.EmptySchema;

namespace Nagasena.Scomp {

  public class EXISchemaFactoryTestUtil {

    private static readonly byte[] m_inputBufferBytes = new byte[1024];

    public static EXISchema getEXISchema() {
      return getEXISchema((string)null, (Object)null);
    }

    internal static EXISchema getEXISchema(string fileName, Object obj) {
      return getEXISchema(fileName, obj, new EXISchemaFactoryTestUtilContext());
    }

    /// <summary>
    /// Loads schema then compiles it into EXISchema.
    /// Schema file is resolved relative to the specified class.
    /// </summary>
    internal static EXISchema getEXISchema(string fileName, Object obj, EXISchemaFactoryTestUtilContext context) {
      if (fileName == null) {
        return EmptySchema.EXISchema;
      }

      Uri uri = getAbsoluteUri(fileName, obj);

      FileStream fos;
      fos = new FileStream(uri.LocalPath, FileMode.Open);

      MemoryStream baos = new MemoryStream();
      int n_bytesRead;
      while ((n_bytesRead = fos.Read(m_inputBufferBytes, 0, m_inputBufferBytes.Length)) != 0) {
        baos.Write(m_inputBufferBytes, 0, n_bytesRead);
      }
      fos.Close();
      byte[] grammarBytes = baos.ToArray();
      baos.Close();

      EXISchema compiled = readEXIGrammar(grammarBytes, context.schemaReader);
      compiled = readEXIGrammar(writeEXIGrammar(compiled, context.stringBuilder), context.schemaReader);
      Stream serialized = serializeSchema(compiled);
      return loadSchema(serialized);
    }

    private static EXISchema loadSchema(Stream @is) {
      EXISchema schema = null;
      try {
        schema = EXISchema.readIn(@is);
      }
      finally {
        @is.Close();
      }
      return schema;
    }

    private static Stream serializeSchema(EXISchema schema) {
      Stream @is = null;
      MemoryStream bos = new MemoryStream();
      try {
        schema.writeOut(bos);
        bos.Flush();
        @is = new MemoryStream(bos.ToArray());
      }
      finally {
        bos.Close();
      }
      return @is;
    }

    /// <summary>
    /// Serialize bytes to a file. </summary>
    /// <param name="bts"> </param>
    /// <param name="baseFileName"> a file that exists in the destination </param>
    /// <param name="fileName"> name of the file the bytes are output to </param>
    /// <param name="cls"> </param>
    public static void serializeBytes(byte[] bts, string baseFileName, string fileName, Object obj) {
      Uri uri = getAbsoluteUri(baseFileName, obj);
      uri = new Uri(uri, fileName);

      FileStream fos;
      fos = new FileStream(uri.LocalPath, FileMode.Create);

      fos.Write(bts, 0, bts.Length);
      fos.Close();
    }

    internal static Uri getAbsoluteUri(String fileName, Object obj) {
      if (obj == null) {
        obj = new EXISchemaFactoryTestUtil();
      }
      String codeBase = obj.GetType().Assembly.CodeBase;
      Uri baseUri = new Uri(codeBase);

      String basePath;
      if (fileName.StartsWith("/")) {
        basePath = "testdata/whatever";
        fileName = fileName.Substring(1);
      }
      else {
        basePath = "testdata/" + obj.GetType().Namespace.Replace(".", "/") + "/whatever";
      }
      baseUri = new Uri(baseUri, basePath);
      return new Uri(baseUri, fileName);
    }

    private static byte[] writeEXIGrammar(EXISchema schema, StringBuilder stringBuilder) {
      MemoryStream outputStream;
      outputStream = new MemoryStream();
      schema.writeXml(outputStream, false);
      byte[] grammarXml = outputStream.ToArray();
      outputStream.Close();
      if (stringBuilder != null) {
        stringBuilder.Remove(0, stringBuilder.Length);
        stringBuilder.Append(System.Text.Encoding.UTF8.GetString(grammarXml));
        //Console.Out.WriteLine(stringBuilder.ToString());
      }
      outputStream = new MemoryStream();
      Transmogrifier transmogrifier = new Transmogrifier();
      transmogrifier.setGrammarCache(GrammarCache4Grammar.GrammarCache, new SchemaId("nagasena:grammar"));
      transmogrifier.OutputOptions = HeaderOptionsOutputType.all;
      transmogrifier.OutputStream = outputStream;
      MemoryStream inputStream = new MemoryStream(grammarXml);
      transmogrifier.encode(new InputSource<Stream>(inputStream));
      inputStream.Close();
      byte[] grammarBytes = outputStream.ToArray();
      outputStream.Close();
      return grammarBytes;
    }

    private static EXISchema readEXIGrammar(byte[] grammarBytes, EXISchemaReader schemaReader) {
      MemoryStream inputStream;
      inputStream = new MemoryStream(grammarBytes);
      if (schemaReader == null) {
        schemaReader = new EXISchemaReader();
      }
      return schemaReader.parse(inputStream);
    }

  }
}